src/ORM/DataList.php

```diff
@@ -1144,139 +1141,157 @@ private function fetchEagerLoadRelations(Query $query): void
-            $joinTableQuery = DB::query('SELECT * FROM "' . $join . '" WHERE "' . $parentField . '" IN (' . implode(',', $ids) . ')');
-            $relationItemIDs = [];
-            $rows = [];
-            while ($row = $joinTableQuery->record()) {
-                $rows[] = [
-                    $parentField => $row[$parentField],
-                    $childField => $row[$childField]
-                ];
-                $relationItemIDs[] = $row[$childField];
+            $joinTableQuery = DB::query('SELECT * FROM "' . $join . '" WHERE "' . $parentIDField . '" IN (' . implode(',', $parentIDs) . ')');
+            $relationItemIDs = $joinTableQuery->column($childIDField);
+            $joinRows = $joinTableQuery;
+        }
+
+        // Get ALL of the items for this relation up front, for ALL of the parents
+        // Fetched as a map so we can get the ID for all records up front (instead of in another nested loop)
+        // Fetched after that as an array because for some reason that performs better in the loop
+        $relationArray = DataObject::get($relationDataClass)->byIDs($relationItemIDs)->map('ID', 'Me')->toArray();
+
+        // Build a map of which children belong to which parent
+        $map = [];
+        foreach ($joinRows as $row) {
+            $parentID = $row[$parentIDField];
+            $childID = $row[$childIDField];
+            if (!isset($map[$parentID])) {
+                $map[$parentID] = [];
             }
+            $map[$parentID][] = $relationArray[$childID];
         }
-        $relationArray = DataObject::get($relationDataClass)->filter(['ID' => $relationItemIDs])->toArray();
-        foreach ($rows as $row) {
-            $eagerLoadID = $row[$parentField];
-            if (!isset($this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relation])) {
-                $arrayList = ArrayList::create();
-                $arrayList->setDataClass($manyManyLastComponent['childClass']);
-                $this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relation] = $arrayList;
+
+        // Store the children in an ArrayList against the correct parent
+        foreach ($map as $parentID => $children) {
+            // There shouldn't be a list for this relation already - if there is, something went wrong.
+            if (isset($this->eagerLoadedData[$eagerLoadRelation][$parentID][$relationName])) {
+                throw new LogicException("Relation list for $relationName on $parentDataClass cannot be fetched twice");
             }
-            $relationItem = array_values(array_filter($relationArray, function ($relationItem) use ($row, $childField) {
-                return $relationItem->ID === $row[$childField];
-            }))[0];
-            $this->eagerLoadedData[$eagerLoadRelation][$eagerLoadID][$relation]->push($relationItem);
+
+            $arrayList = ArrayList::create($children);
+            $this->eagerLoadedData[$eagerLoadRelation][$parentID][$relationName] = $arrayList;
```

Comment:
This whole section of changes is the main performance booster and fixes the root problem in the issue.

It basically boils down to this:
- Previously we were looping through all of the join records, and for every join record we'd then loop through all of the relation records, and call `->ID` on it.
- Now we're getting the IDs up front, and the nested loop is removed entirely.